### ISSUE TITLE
Consider switching from http-parser to llhttp

### ISSUE THREAD

**Issue by PiotrSikora** [envoyproxy/envoy#5155]
Apparently, http_parser is abandonware and node.js started work on a switching to [llhttp](https://llhttp.org/).

cc @alyssawilk @htuch @mattklein123

**Comment by htuch** [comment]
One concern around llhttp is it uses https://github.com/indutny/llparse and a bunch of TypeScript. I don't want to go down the language war path, but it's probably worth considering the increased build complexity (Bazelifying llparser and dependencies) and developer impedance matching when making the call. All things being equal, I'm a fan of higher level solution to parsing, but switching to this is not totally obvious.

**Comment by indutny** [comment]
If I could be of any help, please do not hesitate to ask me directly.

**Comment by alyssawilk** [comment]
@indutny would you mind filing a tracking bug for the migration over on the node.js so interested envoy developers can follow it?  If we opt to migrate we'd probably swap shortly after you, and if we don't it's still important to know when node.js stops using http_parser from a security perspective, so it's good to know either way :-)

**Comment by mattklein123** [comment]
My opinion on this is that we should take a wait and see approach. TBH, HTTP/1.1 at this layer is basically done. I don't really see what changes we would need to make.

**Comment by indutny** [comment]
@alyssawilk https://github.com/nodejs/node/issues/24730 - This might do.

**Comment by htuch** [comment]
Reopening in light of https://github.com/envoyproxy/envoy/issues/6434. Since we didn't get any response from Node.JS security WG when reporting, it seems that http-parser should be considered unsupported. The community did drive a fix when we went public, see https://github.com/nodejs/http-parser/issues/468.

Since http-parser is a codec and hence exposed to untrusted traffic from clients and backends on the data plane, it has a higher standard to live up to than other external dependencies. I would like us to consider only including dependencies that have active security teams that are at least doing CVE management. This is further motivation for moving away from http-parser.

**Comment by indutny** [comment]
@htuch I'm sorry about the negative experience that you had. I wish we would have seen your security report on our channels before you submitted that github issue. What channel did you use? Was it HackerOne or the mailing list (security@nodejs.org)? I'm having a hard time finding your report...

I am part of Node.js community and one of the maintainers of http-parser. I believe that in our shared effort we managed to fix the issue very promptly after it was disclosed. Sorry again if this caused you any frustration!

**Comment by indutny** [comment]
As a matter of fact, I'm a maintainer of llhttp too :joy:

**Comment by htuch** [comment]
@indutny we send two e-mails, one pre and one post disclosure to security@nodejs.org. Happy to forward them again. Is HackeOne preferred?

Thanks again for the prompt fix once this went public, the comment above reflects a concern that we didn't have any effective method to coordinate prior to disclosure.

**Comment by indutny** [comment]
Oh no! security@nodejs.org should be forwarding reports to HackerOne. Apparently, something is not working as expected. Again, I'm terribly sorry for this. Will look into it momentarily.

**Comment by moderation** [comment]
@indutny Will https://hackerone.com/nodejs/hacktivity and https://groups.google.com/group/nodejs-sec be updated retroactively with the CVE?

@htuch it seems like the reporting system would have been or is broken for both http-parser and lhttp (independent of their health it status as sub-projects).

**Comment by indutny** [comment]
@moderation that's a good question. We can definitely request a CVE for the issue. Let me see what I can do.

**Comment by indutny** [comment]
cc @rvagg @jasnell, perhaps?

**Comment by reedloden** [comment]
> @indutny we sent two e-mails, one pre and one post disclosure to [security@nodejs.org](mailto:security@nodejs.org). Happy to forward them again. Is HackerOne preferred? The title was "Vulnerability in http-parser & embedded NULL header handling".

Hi, I work for @Hacker0x01 and act as our main liaison to the open source community. Hoping I can assist you.

When you e-mailed security@nodejs.org, you should have received an e-mail back from HackerOne stating "ACTION REQUIRED: Submit Vulnerability Report". Can you check to see if you received that? If so, you'll need to click that link in order to finalize your report submission.

In any case, you can always submit directly to https://hackerone.com/nodejs.

**Comment by htuch** [comment]
@reedloden as discussed in https://github.com/nodejs/security-wg/issues/454, this was not received. I'm thinking that this is a process flaw if folks can opt to use e-mail or HackerOne and the e-mail path is not as reliable or requires steps that might result in a dropped report. Would it be better to just drop the e-mail option entirely and have folks use HackerOne directly to avoid potential lost reports?

**Comment by jasnell** [comment]
I haven't been as active in the security triage recently but I don't recall this coming through. My apologies for having missed it.

**Comment by reedloden** [comment]
@htuch I figured out what happened, but until it's fixed (either on our side or on Node.js's), I would recommend that https://hackerone.com/nodejs be utilized. Even if e-mail was working, the web interface is preferred, as e-mail is not secure.

**Comment by htuch** [comment]
@reedloden thanks for the analysis of the issue. @indutny are you going to followup with the NodeJS security WG on (1) CVE-2019-9900 and (2) how to deal with the e-mail redirect issues? I can do (1) if that helps by filing a HackerOne report.

**Comment by cmluciano** [comment]
Is [picohttpparser](https://github.com/h2o/picohttpparser) still in the running as a replacement as well? It was mentioned in #2880 & #5536 as a potential perf enhancement for http1.1 .

**Comment by moderation** [comment]
May be time to revisit this. llhttp now the default in Node 12 and the team are looking to deprecate http-parser https://twitter.com/addaleax/status/1173966000175366144

**Comment by htuch** [comment]
Yep, it would be great to see a comparison, including feature support, standard compliance, performance, licenses, security process/maturity and so on. http-parser is basically EOLed.

**Comment by mattklein123** [comment]
Happy to see an analysis done, but IMO we should probably just use llhttp since it's roughly a drop in API replacement for http_parser.

I've used picohttp in the past and we would have to implement a substantial amount of stuff on top that http_parser/llhttp already implements.

**Comment by htuch** [comment]
The only thing that slightly sketches me out about `llhtttp` is the way it pushes us towards having to live in Typescript land. I'm a fan of higher level languages for parsers and codecs, but not such a huge Javascript fan. We can just use the released C implementations though, so build integration shouldn't be an issue. FWIW, the concerns over appropriate handling of vulnerabilities in the Node.js community remains, it took a long time to resolve the http-parser issues I reported.

**Comment by derekargueta** [comment]
Started playing around with llhttp, the parser API is pretty much the exact same which is great, but it doesn't expose the URL parsing library like http_parser which we use in a few places as a URL utility.

**Comment by htuch** [comment]
@derekargueta can we move away from relying on these URL parsing utils and switch to the Chromium URL library as per https://github.com/envoyproxy/envoy/issues/6588? This would allow us to get behind a single source of URL parsing truth, which can only be a good thing for consistency; we've seen quite a few issues with different parts of Envoy not agreeing on things like URL semantics historically.

**Comment by derekargueta** [comment]
@htuch sure, I was unaware of sunsetting the Node URL parser.

Also, should the new parser should initially be gated behind a feature flag to allow opt-in before forcing everywhere, similar to when the buffer implementation changed?

**Comment by htuch** [comment]
@derekargueta I'd recommend a feature flag to allow folks to fall back as needed, but I'd advocate we flip it on eagerly, there isn't a ton of value in remaining with the extant URL parser.

**Comment by derekargueta** [comment]
So far I have 73% of http1 codec tests passing, working through them. llhttp introduces slight deviations - stricter header parsing (i.e. [this block](https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L478) is no longer required), slightly different handling of upgrade, etc.

Also, I'm going to do the URL utility library as a separate change if no one beats me to it. My [updated branch](https://github.com/envoyproxy/envoy/compare/master...derekargueta:dereka/llhttp?expand=1) leaves that as is for now.

**Comment by indutny** [comment]
@derekargueta please let me know if I can be of any help in this endeavor.

**Comment by MylesBorins** [comment]
FWIW I highly suggest moving from http_parser to llhttp. There was even discussion about changing the defaults in 10.x during LTS due to lack of ability to continue to support the deprecated parser

**Comment by mattklein123** [comment]
I will update the title of this issue. We are switching. @derekargueta is working on it.

**Comment by derekargueta** [comment]
This change is about ready, *however* there's one missing feature in llhttp which is the `lenient_http_headers` option (used [here](https://github.com/nodejs/http-parser/blob/master/http_parser.c#L1490) in http-parser). This option toggles strict validation of HTTP header values by ensuring they are printable ASCII characters. We currently run http-parser in lenient mode (which is the default) then provide a runtime switch to enable strict header validation. However, llhttp defaults to strict validation and doesn't provide an option to disable it.

To move forward, we either need to
1. get a lenient option added in llhttp (there's [a TODO for this](https://github.com/nodejs/llhttp/blob/master/src/llhttp/http.ts#L497) cc @indutny) or
2. deprecate `envoy.reloadable_features.strict_header_validation` and have all traffic use strict header value checking. Based on [some conversation](https://github.com/envoyproxy/envoy/pull/7306#discussion_r301618339) on the PR that added strict_header_validation, it seems that we eventually wanted to move this way? If so, I think this option would be better.

**Comment by alyssawilk** [comment]
https://github.com/envoyproxy/envoy/pull/8849 flips strict validation on by default and the option will be removed entirely (unless folks yell) at the next Envoy release.  I assume we're going to support both parsers in parallel for a bit so I think going 100% strict is probably fine (we can revisit if changing the default causes problems in upcoming weeks)

**Comment by mattklein123** [comment]
Awesome stuff. @derekargueta is there any way to support both parsers without a recompile? IMO this should be feature flagged and flippable at startup like we did with the buffer implementation?

**Comment by htuch** [comment]
@derekargueta any update on status here?

**Comment by moderation** [comment]
@htuch I've done my part to encourage @derekargueta :wink: https://twitter.com/moderat10n/status/1230125590428053509

**Comment by derekargueta** [comment]
No status update at the moment - I have started catching my stale branches up to master. There's been some refactorings of the HTTP codecs that I need to work through

**Comment by derekargueta** [comment]
Started re-investigating this. @indutny one issue that came up is that the llhttp release artifact is that it's not built with `strict` mode enabled, and this option cannot be enabled from the C API since it is a variable in producing the C code. Are there plans to publish release builds with strict mode enabled, or be able to set strict mode through the C-API perhaps using `-D` like http-parser? This is causing quite a few tests to fail but I really don't want to introduce Typescript tooling in the Envoy build (for running `npx` to generate the C code with strict mode)

**Comment by indutny** [comment]
There is a bit of optimizing that has to be done on llhttp's code before compiling it down to C. I don't think that it'd be possible to introduce the configuration defines the same way as they are in http-parser. Runtime checks would likely affect performance so I'd avoid this too.

There are two things that could be done, however:

* Bundling two separate builds and switching between them in a build system (Easy, might be aesthetically undesirable)
* Combining two generated `.c` files from strict and loose mode builds into one guarded by a huge top-level `#ifdef`

I'll look into the latter to see how much churn this would involve in llhttp. Could you consider the former meanwhile?

**Comment by indutny** [comment]
See: https://github.com/nodejs/llhttp/pull/58

**Comment by indutny** [comment]
FWIW, that Pull Request has landed and was released in llhttp 2.1.0.

**Comment by htuch** [comment]
@derekargueta any update on progress here?

**Comment by moderation** [comment]
Now @derekargueta 's [presentation](https://kccnceu20.sched.com/event/ZfGp/keynote-building-a-service-mesh-from-scratch-the-pinterest-story-derek-argueta-senior-software-engineer-tesla) at KubeConEU is done he should have some cycles :smile:

**Comment by moderation** [comment]
This commit on 2020-10-02 confirms that the last remaining maintainer for http-parser has moved on and http-parser is no longer maintained - https://github.com/nodejs/http-parser/commit/ec8b5ee63f0e51191ea43bb0c6eac7bfbff3141d.

From an @envoyproxy/dependency-shepherds POV it would be great if we could get https://github.com/envoyproxy/envoy/pull/9033 over the line /cc @derekargueta @htuch

**Comment by htuch** [comment]
I've tagged this as both priority and security in response to the above comment.

**Comment by derekargueta** [comment]
I have a ["drop-in replacement" branch](https://github.com/envoyproxy/envoy/compare/master...derekargueta:llhttp-swap-in) with everything passing _except_ the new functionality that was introduced via http-parser for allowing `transfer-encoding: chunked` + `content-length: N`. There's an llhttp issue for this where I've started a small investigation and the fix seems pretty straightforward: https://github.com/nodejs/llhttp/issues/69

**Comment by pallas** [comment]
FYI, nodejs/llhttp#69 was just closed with a fix in v3.0.0

**Comment by moderation** [comment]
I added a link to the release at the now stale WIP PR at https://github.com/envoyproxy/envoy/pull/9033#discussion_r532955677

**Comment by derekargueta** [comment]
Confirmed all tests pass with the new llhttp update 🎉 reviving the abstractions I initially had for supporting both parsers through a command line flag

**Comment by moderation** [comment]
Woohoo

**Comment by asraa** [comment]
I ran some perf tests and got CPU profiles from Derek's branches (direct swap in with llhttp, and an added parser factory and parser class) here: https://gist.github.com/asraa/8babd3516acd447aa0f2ff5aa61d2e19

It doesn't seem to me that there's significant overhead in adding the factory, and llhttp swap in results in some CPU time differences (IMO insignificant).

If there's anything I can do to make some of these benchmarks more informative please let me know!

**Comment by htuch** [comment]
@asraa looks compelling, i.e. there is little difference. I think the 1M QPS example probably has a ton of error rate reflecting open loop request drops; it might be a good idea when measuring max QPS achievable to switch NH to closed loop mode (at which point latency is not very interesting). I'd also be curious to see if there is any difference for large request, e.g. 1MB body.

**Comment by asraa** [comment]
Thanks! More benchmarks added.
Those benchmarks were actually closed loop, that appears the default in the nighthawk test client. I ran with `--open-loop` again though, and added those benchmarks. Again, didn't see a major difference in open loop latency. 

You do see a larger error rate (OR llhttp improvement) in the open loop 500k QPS (1M was enough to break nighthawks stats gathering).

**Comment by asraa** [comment]
BTW -- added an adaptive load test (with nighthawk) to find optimal RPS (based on mean latency <= 5ms and 95%+ success rate)
@derekargueta 

| - | BASELINE http_parser | llhttp | Factory with http_parser | Factory with llhttp |
| --- | --- | ---| --- | --- |
| Final RPS | 20125 | 20156 | 19531 | 19843 |

the llhttp boost somewhat counteracts the factory, there's a 1% degradation.

**Comment by moderation** [comment]
llhttp 4.0.0 is out - https://github.com/nodejs/llhttp/releases/tag/release%2Fv4.0.0

**Comment by moderation** [comment]
https://nodejs.org/en/blog/release/v12.22.0/
> The legacy HTTP parser is runtime deprecated

> The legacy HTTP parser, selected by the --http-parser=legacy command line option, is deprecated with the pending End-of-Life of Node.js 10.x (where it is the only HTTP parser implementation provided) at the end of April 2021. It will now warn on use but otherwise continue to function and may be removed in a future Node.js 12.x release.

> The default HTTP parser based on llhttp is not affected. By default it is stricter than the now deprecated legacy HTTP parser. If interoperability with HTTP implementations that send invalid HTTP headers is required, the HTTP parser can be started in a less secure mode with the --insecure-http-parser command line option.

**Comment by klarose** [comment]
@asraa Do you know offhand whether or not this new parser implementation will allow Envoy to support custom methods for http 1.1? I'm having to bypass the http layer of envoy in my cluster for a service whose protocol (over which I have no control) uses non-standard http methods. That is quite unfortunate.

I took a quick gander at the llhttp code, and it looked like it may have a similar restriction to the old parser (i.e. a see explicit lists of http methods e.g. https://github.com/nodejs/llhttp/blob/master/src/llhttp/c-headers.ts#L46), which worries me. I could be wrong, though.

**Comment by kanongil** [comment]
@klarose Last I checked llhttp hardcodes the same whitelisted HTTP methods.

**Comment by klarose** [comment]
> @klarose Last I checked llhttp hardcodes the same whitelisted HTTP methods.

Thanks for looking in to it. That's unfortunate. Do you know if there is any way to change that behaviour? It's quite awful for us. We don't have control over the protocol involved here. Having to treat the request as TCP because the envoy http parser cannot handle unknown methods really removes a bunch of its benefits -- we lose the logging, filters, etc that all come with http. All of those things support arbitrary methods. Why not the parser?

There is some discussion here about the meaning of the methods, as well as a registry: https://datatracker.ietf.org/doc/html/rfc7231#section-4. It does say that a method which cannot be handled by a server should be rejected. However, envoy is not an http server; it is a proxy.

It seems to me that the server to which envoy is proxying should utltimately decide whether a method is unsupported. 

Perhaps this isn't the proper forum for discussing this, but I feel like it should be taken into consideration when decision to use this parser.

**Comment by alyssawilk** [comment]
Google has a similar problem where we need to support methods which http_parser does not handle.   cc @yanavlasov maybe link your doc on this thread?

**Comment by alyssawilk** [comment]
Closing in favor of https://github.com/envoyproxy/envoy/issues/21245
